src/rez/package_order.py

```diff
@@ -1,36 +1,46 @@
 # SPDX-License-Identifier: Apache-2.0
 # Copyright Contributors to the Rez Project
-
+from __future__ import annotations
 
 from inspect import isclass
 from hashlib import sha1
-from typing import Dict, Iterable, List, Optional, Union
+from typing import Any, Callable, Dict, Iterable, Self
+from typing_extensions import Protocol
 
 from rez.config import config
 from rez.utils.data_utils import cached_class_property
 from rez.version import Version, VersionRange
 from rez.version._version import _Comparable, _ReversedComparable, _LowerBound, _UpperBound, _Bound
-from rez.packages import iter_packages
+from rez.packages import iter_packages, Package
 
 ALL_PACKAGES = "*"
 
 
+class SupportsLessThan(Protocol):
+    def __lt__(self, __other: Any) -> bool: ...
+
+
 class FallbackComparable(_Comparable):
     """First tries to compare objects using the main_comparable, but if that
     fails, compares using the fallback_comparable object.
     """
 
-    def __init__(self, main_comparable, fallback_comparable):
+    def __init__(self, main_comparable: SupportsLessThan,
+                 fallback_comparable: SupportsLessThan):
         self.main_comparable = main_comparable
         self.fallback_comparable = fallback_comparable
 
-    def __eq__(self, other):
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, FallbackComparable):
+            return NotImplemented
```

Comment:
`mypy` strongly prefers that comparison functions define the type of `other` as `object`. by returning `NotImplemented` we allow other other object to handle equality if it implements it.   This should not change the behavior as long as `other` is always `FallbackComparable` within Rez.

Reply:
will do.